server/src/main/java/io/deephaven/server/plugin/ObjectTypes.java

```diff
@@ -0,0 +1,65 @@
+package io.deephaven.server.plugin;
+
+import io.deephaven.plugin.type.ObjectType;
+import io.deephaven.plugin.type.ObjectTypeClassBase;
+import io.deephaven.plugin.type.ObjectTypeLookup;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+final class ObjectTypes implements ObjectTypeLookup {
+
+    private static final Set<String> RESERVED_TYPE_NAMES = Set.of("Table", "TableMap", "TreeTable");
+
+    private final Set<String> names;
+    private final Map<Class<?>, ObjectType> classTypes;
+    private final List<ObjectType> otherTypes;
+
+    public ObjectTypes() {
+        names = new HashSet<>();
+        classTypes = new HashMap<>();
+        otherTypes = new ArrayList<>();
+    }
+
+    @Override
+    public synchronized Optional<ObjectType> findObjectType(Object object) {
+        final ObjectType byClass = classTypes.get(object.getClass());
+        if (byClass != null) {
+            return Optional.of(byClass);
+        }
+        for (ObjectType type : otherTypes) {
+            if (type.isType(object)) {
+                return Optional.of(type);
+            }
+        }
+        return Optional.empty();
+    }
+
+    public synchronized void register(ObjectType type) {
```

Comment:
Right now, I'm thinking that it's only via plugin mechanisms. We could expose the `register`, but I *think* it has some non-obvious implications for the script session (and application fields) `changeListener` wrt "displayble" types.

Reply:
I'm definitely imagining a use case where a java/groovy user would want to iterate without knocking over the server with every change - just keep calling register (and a matching unregister?) as you tweak the class definition in the console, fetch from the client to work on it. Probably matters less locally, but maybe more on a server deployment. 

This might just be in my head - but with such a flexible mechanism (and already threadsafe etc), it seems like a shame to keep it from power users.

Reply:
punting for now https://github.com/deephaven/deephaven-core/issues/1809